<!-- JPN -->
# NumPyによる線形代数

※本演習資料の二次配布・再配布はお断り致します。

<!-- ENG -->
# Linear algebra using NumPy

※Distribution or redistribution of these exercise materials without the copyright holder's permission is not permitted.

<!-- JPN -->
　本日は、講義で機械学習等で用いることの多い線形代数を学んだ。演習では、**NumPy**と呼ばれるPythonライブラリ（便利なツールセットのようなもの）について学習し、データの処理や講義で学んだ固有値分解などを実際に行ってみることにする。

　併せて、講義において学んだ「数式的な」結果とは異なり、コンピュータが一般的に行う「数値的な」結果には微妙な誤差が含まれることも、ここで確認していきたい。

<!-- ENG -->
　This time, in the lecture, we learned about linear algebra, which is often used in machine learning and other applications. In the exercises, we will learn about a Python library called **NumPy** (a useful toolset) and try to process data and perform processings such as eigendecomposition, which we learned in the lecture.

　At the same time, we would like to confirm that, unlike the "mathematical" results we learned in the lecture, the "numerical" results that computers generally produce contain subtle errors.

<!-- JPN -->
## 1 | NumPy入門

　ベクトルや行列などの表現をはじめとして、Pythonでデータを処理するときに極めてよく使われるのがNumPyである。ここでは、基盤人工知能演習・基盤データサイエンス演習で必要になりうる機能や関数に絞って、使い方を紹介する。より詳細なチュートリアルは以下の資料を見ると良い。

* [Chainer NumPy入門（日本語）](https://tutorials.chainer.org/ja/08_Introduction_to_NumPy.html)
* [NumPy公式チュートリアル（英語）](https://docs.scipy.org/doc/numpy/user/quickstart.html)



<!-- ENG -->
## 1 | Introduction to NumPy

　NumPy is commonly used all the time to process data in Python, including vectors, matrices, and other representations. In this section, we will focus on functions and features that may be required for the Exercise in Fundamentals of Artificial Intelligence / Exercise in Fundamentals of Data Science, as well as introduce their usage. A more detailed tutorial can be found in the following materials:

* [Intro to NumPy in Chainer Tutorial (Japanese)](https://tutorials.chainer.org/ja/08_Introduction_to_NumPy.html)
* [NumPy quickstart (English)](https://docs.scipy.org/doc/numpy/user/quickstart.html)



<!-- JPN -->
　NumPyのデータ構造や関数（メソッド）を用いる場合は、最初に以下のimport文を書く必要がある。`np`はそれ以外の名称にすることも可能だが、慣例的に`np`とされている。

<!-- ENG -->
　When using NumPy data structures and functions (methods), the following import statement must be written first. The `np` can be called something else, but it is conventionally called `np`.

In [2]:
import numpy as np # It means you use NumPy in the name of np

<!-- JPN -->
### 1.1 | NumPy配列

　NumPyの根幹をなすものがNumPyの配列 `np.array` である。 `np.array` はPythonの最初から準備されている配列とは大きく異なり、様々な便利な挙動を示す（補足資料 **※1**）。

<!-- ENG -->
### 1.1 | NumPy arrays

　The basis of NumPy is the NumPy arrays, `np.array`. `np.array` is very different from the built-in arrays in Python, and shows various useful behaviors (Supplementary Material **S1**).

<!-- JPN -->
　まずは、NumPy配列を定義してみる。 `np.array()` という関数のようなものの引数にpythonの配列を入れることで作ることができる（補足資料 **※2**）。

<!-- ENG -->
　First, we try to define a NumPy array. It can be created by putting a Python array in the argument of something like a function called `np.array()` (Supplementary Material **S2**).

In [3]:
arr = np.array([1, 2, 3])
print(arr)
print(arr.shape) # Check the shape of arr

[1 2 3]
(3,)


<!-- JPN -->
　同様に、行列のような縦横二次元の配列も表現することができる。

<!-- ENG -->
　Similarly, a two-dimensional array such as a matrix can be represented.

In [4]:
arr = np.array([[1, 2],
                [3, 4],
                [5, 6]])
print(arr)
print(arr.shape) # Check the shape of arr

[[1 2]
 [3 4]
 [5 6]]
(3, 2)


<!-- JPN -->
　次に、arr同士の加減乗除を行ってみる。Pythonの配列とは全く異なる挙動を示すことを確認してほしい。

<!-- ENG -->
　Next, we try to add, subtract, multiply and divide the arrays themselves. You should see that it behaves quite differently from Python arrays.

In [5]:
arr1 = np.array([1,2,3])
arr2 = np.array([4,5,6])
print(arr1 + arr2)
print(arr1 - arr2)
print(arr1 * arr2)
print(arr1 / arr2)

[5 7 9]
[-3 -3 -3]
[ 4 10 18]
[0.25 0.4  0.5 ]


<!-- JPN -->
　上のセルの結果を見ると、2つの `np.array` 、 `arr1` と `arr2` の加減乗除はいずれも要素ごとの計算が行われている。

<!-- ENG -->
　Looking at the results in the cell above, addition, subtraction, multiplication, and division of the two `np.array`, `arr1` and `arr2`, are all calculated element by element.

<!-- JPN -->
　なお、この **`np.array` の挙動はPythonに元から定義されているリスト `list` とは大きく異なっている**。以下に算術演算子を用いた結果をまとめて記述する。

| 演算 | `list` | `np.array` |
| ---- | ------ | ---------- |
| `+`  | [1, 2, 3, 4, 5, 6] | [5, 7, 9] |
| `-`  | （エラー） | [-3, -3, -3] |
| `*`  | （エラー） | [4, 10, 18] |
| `/`  | （エラー） | [0.25, 0.4, 0.5] |


<!-- ENG -->
　Note that **the behavior of `np.array` is very different from the Python `list`**. The following is a summary of the results of using arithmetic operators.

| Operators | `list` | `np.array` |
| ---- | ------ | ---------- |
| `+`  | [1, 2, 3, 4, 5, 6] | [5, 7, 9] |
| `-`  | (Error) | [-3, -3, -3] |
| `*`  | (Error) | [4, 10, 18] |
| `/`  | (Error) | [0.25, 0.4, 0.5] |


<!-- JPN -->
　もちろん、 `np.array` で定義された二次元の配列についても同様に要素ごとの加減乗除が行われる。

<!-- ENG -->
　Of course, the two dimensional array defined in `np.array` is also added, subtracted, multiplied, and divided for each element.

In [6]:
arr1 = np.array([[1,2,3],
                 [4,5,6]])
arr2 = np.array([[2,3,4],
                 [5,6,7]])
print(arr1 + arr2)
print(arr1 - arr2)
print(arr1 * arr2)
print(arr1 / arr2)

[[ 3  5  7]
 [ 9 11 13]]
[[-1 -1 -1]
 [-1 -1 -1]]
[[ 2  6 12]
 [20 30 42]]
[[0.5        0.66666667 0.75      ]
 [0.8        0.83333333 0.85714286]]


<!-- JPN -->
　次に、 `np.array` とスカラー値との四則演算を行ってみよう。

<!-- ENG -->
　Next, let's perform four arithmetic operations on `np.array` with scalar values.

In [7]:
arr1 = np.array([1,2,3])
print(arr1 + 2)
print(arr1 - 2)
print(arr1 * 2)
print(arr1 / 2)

[3 4 5]
[-1  0  1]
[2 4 6]
[0.5 1.  1.5]


In [8]:
arr2 = np.array([[1,2,3],
                 [4,5,6]])
print(arr2 + 2)
print(arr2 - 2)
print(arr2 * 2)
print(arr2 / 2)

[[3 4 5]
 [6 7 8]]
[[-1  0  1]
 [ 2  3  4]]
[[ 2  4  6]
 [ 8 10 12]]
[[0.5 1.  1.5]
 [2.  2.5 3. ]]


<!-- JPN -->
　このように、各要素に対する計算が行われる。

<!-- ENG -->
　In this way, calculations for each element are performed.

<!-- JPN -->
### 1.2 | `np.array` 内の要素の取得と編集



<!-- ENG -->
### 1.2 | Getting and editing elements from `np.array`



<!-- JPN -->
　`np.array` の各要素は、以下のように1つ1つの値を出力したり、書き換えることが可能である。

<!-- ENG -->
　Each `np.array` element can be output or replaced one by one as follows.

In [9]:
arr = np.array([[1,2], [3,4]])
print(arr)

[[1 2]
 [3 4]]


In [10]:
print(arr[1,1]) # It extracts the element at row 1 and column 1 (it means SECOND row, SECOND column) by putting values separated by commas.
                # Note that the top-left element is the 0th position.

4


In [11]:
arr[1,0] = 100 # Replace a value
print(arr)     # Check the replacement has been reflected

[[  1   2]
 [100   4]]


<!-- JPN -->
　また、ある1列、1行だけを取り出したい場合にはコロン <code>:</code>を利用する。

<!-- ENG -->
　Also, if you want to extract only one column or one row, use a colon <code>:</code>.

In [12]:
print(arr[0, :]) # Extract row 0 (note that it will be a one dimensional array, not a two dimensional array)
print(arr[:, 0]) # Extract column 0 (note that it will be a one dimensional array, not a two dimensional array)

[1 2]
[  1 100]


<!-- JPN -->
### 1.3 | 1次元配列の列（縦）ベクトル化、行（横）ベクトル化

　`np.array`の1次元配列を列ベクトル、あるいは行ベクトルに書き換える必要が時々発生する（詳細は後述するが、 `np.array` の1次元配列は列ベクトルとも行ベクトルともつかない、難解な挙動を示す）。その場合は、以下のように `reshape()` を利用すると良い。

<!-- ENG -->
### 1.3 | One dimensional array column (vertical) vectorization, row (horizontal) vectorization

　It is sometimes necessary to reshape `np.array` one dimensional arrays as column vectors or row vectors (as we will see in detail later, `np.array` one dimensional arrays can show a strange behavior that is neither a column vector nor a row vector). In such a case, you can use `reshape()` as follows.

In [13]:
arr = np.array([1, 2, 3])
v_vec = arr.reshape(-1, 1) # Make it a column vector (a matrix of n rows and one column), setting the n part to -1.
print(v_vec)

[[1]
 [2]
 [3]]


In [14]:
arr = np.array([4, 5, 6])
h_vec = arr.reshape(1, -1) # Make it a row vector (a matrix of 1 row and n columns), setting the n part to -1.
print(h_vec)

[[4 5 6]]


<!-- JPN -->
### 1.4 | 行列と行列、行列とベクトル、行列と1次元配列の結合

　`np.concatenate()`, `np.vstack()`, `np.hstack()` を使うことで2つ以上の行列の結合を行うことができる。

<!-- ENG -->
### 1.4 |  Combining matrices and matrices, matrices and vectors, matrices and one dimensional arrays

　Two or more matrices can be combined by using `np.concatenate()`, `np.vstack()`, and `np.hstack()`.

In [15]:
mat1 = np.array([[1, 2],
                 [3, 4]])
mat2 = np.array([[5, 6],
                 [7, 8]])
print(np.vstack([mat1, mat2]))              # Combine vertically
print(np.concatenate([mat1, mat2], axis=0)) # Combine vertically, the same as np.vstack

print(np.hstack([mat1, mat2]))              # Combine horizontally
print(np.concatenate([mat1, mat2], axis=1)) # Combine horizontally, the same as np.hstack


[[1 2]
 [3 4]
 [5 6]
 [7 8]]
[[1 2]
 [3 4]
 [5 6]
 [7 8]]
[[1 2 5 6]
 [3 4 7 8]]
[[1 2 5 6]
 [3 4 7 8]]


<img src="https://i.imgur.com/Qwz4EaW.png" width=500>

<!-- JPN -->
　**`np.concatenate` や `np.vstack`, `np.hstack`は2次元配列同士にしか適用できない**ため、行列と1次元配列を結合する場合は 1.3節 の方法を用いて列ベクトルか行ベクトルに変換する必要がある。行列や、結合する方向を意識して、列ベクトルにするか行ベクトルにするか考えて結合させよう。

<!-- ENG -->
　Since **`np.concatenate`, `np.vstack`, and `np.hstack` can only be applied to between two dimensional arrays**, if you want to combine a matrix with a one dimensional array, you need to convert it to a column vector or a row vector using the method indicated in section 1.3. You should combine them together thinking whether to make it a column vector or a row vector. Be aware of the matrix shape and the orientation when combining.

In [16]:
mat = np.array([[1,2],
                [3,4]])
arr = np.array([5,6])
h_vec = arr.reshape(1, -1)
v_vec = arr.reshape(-1, 1)

print(np.vstack([mat, h_vec])) # Combine vertically
print(np.hstack([mat, v_vec])) # Combine horizontally

[[1 2]
 [3 4]
 [5 6]]
[[1 2 5]
 [3 4 6]]


<!-- JPN -->
### 1.5 | 行列の列単位・行単位の計算

　Pythonにおける機械学習では、1行ごとに1つのデータ（例：英語・数学・化学・物理の得点）を置くことで行列を構成し、その行列に対する演算を行うようなことが多々発生する。そんなときに便利な機能をいくつか説明していく。

<!-- ENG -->
### 1.5 | Column-wise and row-wise calculation of matrices

　Machine learning in Python often involves constructing matrices by putting one item of data per row (e.g., English, math, chemistry, and physics scores) and performing operations on the matrices. We will see some of the functions that are useful in such cases.

In [17]:
# An example of data set
scores = np.array([[50, 50, 60, 50],  # The first person's scores in English, math, chemistry, and physics
                   [40, 80, 20, 10],  # The second person's scores in English, math, chemistry, and physics
                   [90, 20, 40, 40]]) # The third person's scores

<!-- JPN -->
#### 列方向・行方向の集計

　例えば、上記のような試験結果（＝サンプル）があるときに、以下のような値を求めたいとする。

* 各教科の平均点
* 各学生の合計点

　これらを算出するには、行列の各列、あるいは各行の総和や平均値をとる必要があるが、これは以下のように計算を行うことができる。

<!-- ENG -->
#### Column and row calculation

　For example, suppose you want to find the following values ​​when you have the above test results (= sample data).

* Average score for each subject
* Total score for each student

　To calculate these, it is necessary to take the sum or average value of each column or row of the matrix, which can be calculated as follows.

In [18]:
# Average score for each subject
print(scores.mean(axis=0)) # Find the average of each column
# print(np.mean(scores, axis=0)) # Can do the same with this code

[60.         50.         40.         33.33333333]


In [19]:
# Total score for each student
print(scores.sum(axis=1)) # Find the sum of each row
# print(np.sum(scores, axis=1)) # Can do the same with this code

[210 150 190]


<img src="https://i.imgur.com/STyCbNN.png" width=500>

<!-- JPN -->
　このように、`axis`を指定することで、各列、各行の集計を簡単に行うことができる。

<!-- ENG -->
　In this way, you can easily calculate each column and each row by specifying the `axis`.

<!-- JPN -->
#### ブロードキャスト：各列・各行への算術演算

　次に、それぞれの学生の各教科の得点が、各教科の平均点からどれほど高いのか、あるいは低いのか計算したい場合を考える。

　まず各教科の平均点を計算する。

<!-- ENG -->
#### Broadcasting: arithmetic operation for each column and row

　Next, consider the case where you want to calculate how much higher or lower each student's score in each subject is from the average score in each subject.

　First, calculate the average score for each subject.

In [20]:
averages = scores.mean(axis=0)

<!-- JPN -->
　次に、それぞれの学生の各教科の得点と、平均点との差を計算するのだが、これは以下のように簡単に計算することができる。

<!-- ENG -->
　The next step is to calculate the difference between each student's score in each subject and the average score, which can be easily calculated as follows.

In [21]:
print(scores - averages)

[[-10.           0.          20.          16.66666667]
 [-20.          30.         -20.         -23.33333333]
 [ 30.         -30.           0.           6.66666667]]


<!-- JPN -->
　この引き算は行列 `scores` と 1次元配列 `averages` との間で行われていることに注意しなければならない。
NumPyでは**行列の列数と1次元配列の要素数が一致している場合**に限り、**行列の各行に対して**1次元配列との算術演算が行える（**補足資料 ※3**）。これを**ブロードキャスト**と言う。

　ブロードキャストは極めて便利である一方、**行列、ベクトル、1次元配列のサイズの意識がおろそかになり、行と列を間違えた計算を引き起こす原因にもなる**。行列などのサイズを意識し、**想定通りの結果になっているかどうか、適宜 `print()` を行いながらコードを記述することが大切**である。

<!-- ENG -->
　Note that this subtraction is performed between the matrix `scores` and the one dimensional array `averages`.
In NumPy, arithmetic operations with one dimensional arrays can be performed **for each row of a matrix** only **when the number of columns of the matrix and the number of elements of the one dimensional array are the same** (**Supplementary Material S3**). This is called **broadcasting**.

　While broadcasting is extremely convenient, **it leads to carelessness of the size of matrices, vectors, and one dimensional arrays, and can cause miscalculations due to confusion between rows and columns**. It is important to be aware of the size of matrices, etc., and **to write the code while performing `print()` as appropriate to see if the results are as expected**.

<img src="https://i.imgur.com/gkwvtJ3.png" width=500>

<!-- JPN -->
　例えば、以下のような**axis指定を忘れた**というコードミスを行ってみよう。`mean()`でaxis指定を忘れると、**行列内の全要素の平均値**を算出してしまう。

<!-- ENG -->
　For example, try to make a coding mistake of **forgetting to specify the axis** as follows. If you forget to specify the axis in `mean()`, it will calculate **the average value of all elements in the matrix**.

In [22]:
# Try forgetting to specify the axis
averages = scores.mean() # Oh!
print(scores - averages) # No errors were encountered, however, it was not the answer we had expected

[[  4.16666667   4.16666667  14.16666667   4.16666667]
 [ -5.83333333  34.16666667 -25.83333333 -35.83333333]
 [ 44.16666667 -25.83333333  -5.83333333  -5.83333333]]


<!-- JPN -->
　axis指定を行わない `mean()` は全要素の平均を算出するので、スカラー値になる。行列とスカラー値の減算は第1章で説明した通りに実行できてしまうので、**エラーは発生しない**。しかし、その結果は各教科の平均点からの差分、という当初の目的とは異なったものになっている。このように**「文法は合っているが、論理的には間違っている」バグは発見が難しい**ので、先ほど述べたように `print()` を行うなどして間違っていないか確認するようにしよう。

<!-- ENG -->
　`mean()` without an axis specification calculates the average of all elements, so it will be a scalar value. The subtraction of matrices and scalar values can be performed as described in 1, so **no error will occur**. However, the results are different from the original purpose of our exercise, which is to find the difference from the average score of each subject. **This kind of "syntax correct, but logically wrong" bug is difficult to detect**, so make sure to check for it by using `print()` as mentioned earlier.

In [23]:
print(scores.mean()) # Huh, we're only getting one value, not what we were expecting. It's a bug!

45.833333333333336


-----

<!-- JPN -->
##### 課題 1

　上記 `scores` を用いて、各教科の平均点 `averages` 、各教科の標準偏差 `stds` 、および3人の各教科の偏差値 `t_scores` を算出するコードを作成せよ。

<!-- ENG -->
##### Exercise 1

　Using the above `scores`, calculate the average score of each subject `averages`, the standard deviation of each subject `stds`, and the deviation in each subject for these three students `t_scores`.

<!-- JPN -->
　なお、学生 $i$ の教科 $x$ における偏差値は以下の計算式で計算できる。$x_i$ は学生 $i$ の 教科 $x$ の得点、$\mu_x, \sigma_x$は教科 $x$ の平均点と標準偏差である。

<!-- ENG -->
　Note that the deviation of student $i$ in subject $x$ can be calculated using the following formula. $x_i$ is the score of student $i$ in subject $x$, and $\mu_x, \sigma_x$ are the mean score and standard deviation of subject $x$.

<!-- BOTH -->
$$
T_{i, x} = \frac{10(x_i - \mu_x)}{\sigma_x} + 50
$$

<!-- JPN -->
また、標準偏差の計算には標本分散を利用せよ。

<!-- ENG -->
Use sample variance for the calculation of standard deviation.

In [27]:
# Write your answers here
import numpy as np

scores = np.array([[50, 50, 60, 50],  # The first person's scores in English, math, chemistry, and physics
                   [40, 80, 20, 10],  # The second person's scores in English, math, chemistry, and physics
                   [90, 20, 40, 40]]) # The third person's scores

averages = np.mean(scores, axis=0)
stds     = np.std(scores, axis=0, ddof=1)
t_scores = 10 * ((scores - averages) / stds) + 50

print("Averages:", averages)
print("Standard deviations:", stds)
print("T-scores:\n", t_scores)

Averages: [60.         50.         40.         33.33333333]
Standard deviations: [26.45751311 30.         20.         20.81665999]
T-scores:
 [[46.22035527 50.         60.         58.00640769]
 [42.44071054 60.         40.         38.79102923]
 [61.33893419 40.         50.         53.20256308]]


-----

<!-- JPN -->
## 2 | NumPyによる線形代数

<!-- ENG -->
## 2 | Linear algebra using NumPy

<!-- JPN -->
### 2.1 | ベクトルと行列

　ここまで、 `np.array` のデータの取り扱いや四則演算に関する挙動を見てきたが、線形代数では例えば**行列積の計算**や、**行列とベクトルの積**、あるいは**ベクトル同士の内積**等を行う必要がある。

　なお、**線形代数におけるベクトルを定義する際には、列ベクトルか行ベクトルかを明示的にするために、2次元配列として定義してほしい**（例えば、列ベクトルはn行1列の行列として扱う、など）。



<!-- ENG -->
### 2.1 | Vectors and matrices

　So far, we have seen the handling of `np.array` data and the behavior related to the four arithmetic operations, but in linear algebra, for example, it is necessary to calculate **the matrix product**, **the product of the matrix and the vector**, or **the internal product of the vectors**.

　**When defining vectors in linear algebra, please define them as two dimensional arrays in order to make it explicit whether they are column vectors or row vectors** (For example, a column vector is treated as a matrix of n rows and one column, etc.).



In [28]:
vector = np.array([[1],
                   [2],
                   [3]]) # Column vector consisting of three elements
matrix = np.array([[1, 2],
                   [3, 4],
                   [5, 6]])  # A matrix with three rows and two columns
print("print(vector)")
print(vector)
print("print(matrix)")
print(matrix)

print(vector)
[[1]
 [2]
 [3]]
print(matrix)
[[1 2]
 [3 4]
 [5 6]]


<!-- JPN -->
　作成したベクトルや行列の形状は`vector.shape`、`matrix.shape`で確認可能である。

<!-- ENG -->
　The shapes of the created vectors and matrices can be checked in `vector.shape` and `matrix.shape`.

In [29]:
print(vector.shape) # Column vector with three elements -> (3, 1)
print(matrix.shape) # A matrix with three rows and two columns  -> (3, 2)

(3, 1)
(3, 2)


<!-- JPN -->
### 2.2 | 行列・ベクトルの転置

　行列の転置は非常に簡単であり、 `mat.T` `vec.T` とすれば転置した行列を得ることができる。


<!-- ENG -->
### 2.2 | Transposition of matrices and vectors

　The transposition of the matrix is very simple, and we can get the transposed matrix by setting `mat.T` or `vec.T`.


In [30]:
mat = np.array([[1,2],
                [3,4]])
print(mat)
print(mat.T)

[[1 2]
 [3 4]]
[[1 3]
 [2 4]]


<!-- JPN -->
　正しく、$\mathrm{mat}^\mathrm{T} = \begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}^\mathrm{T} = \begin{bmatrix}
1 & 3 \\
2 & 4 \\
\end{bmatrix}$ となっている。

<!-- ENG -->
　See that we have a correct output: $\mathrm{mat}^\mathrm{T} = \begin{bmatrix}
1 & 2 \\
3 & 4 \\
\end{bmatrix}^\mathrm{T} = \begin{bmatrix}
1 & 3 \\
2 & 4 \\
\end{bmatrix}$.

<!-- JPN -->
### 2.3 | 行列積、行列とベクトルの積、ベクトル間の内積
　1.1節で説明したように、2次元配列の`np.array`同士の積を `mat * mat` と記述しても、要素ごとの積を行ってしまい、**行列積を行うことはできない**。

<!-- ENG -->
### 2.3 | Matrix product, product of a matrix and a vector, inner product between vectors
　As explained in Section 1.1, writing `mat * mat` as a product between `np.arrays` of two dimensional arrays will calculate an element by element product and **not a matrix product**.

In [31]:
mat1 = np.array([[1, 2],
                 [3, 4]])
mat2 = np.array([[5, 6],
                 [7, 8]])
print(mat1 * mat2) # Check the results

[[ 5 12]
 [21 32]]


<!-- JPN -->
　行列積を行う場合には、以下のように、`np.dot()` あるいは`arr.dot()` を利用する（`mat1 @ mat2` としても計算できるが、直感的ではないので本演習では非推奨とする）。

<!-- ENG -->
　To do matrix products, use `np.dot()` or `arr.dot()` as follows (it can also be calculated as `mat1 @ mat2`, but not recommended for this exercise as it is not intuitive).

In [32]:
print(mat1.dot(mat2))     # Matrix product of mat1 and mat2
print(np.dot(mat1, mat2)) # Same meaning as above

[[19 22]
 [43 50]]
[[19 22]
 [43 50]]


In [33]:
print(mat2.dot(mat1))      # The product of a matrix depends on the order
print(np.dot(mat2, mat1))

[[23 34]
 [31 46]]
[[23 34]
 [31 46]]


<!-- JPN -->
　同様に、行列とベクトルの積、ベクトル間の内積も `dot` を利用する。

<!-- ENG -->
　In the same way, products of matrices and vectors and inner products between vectors also use `dot`.

In [34]:
mat = np.array([[1,2],
                [3,4],
                [5,6]])
vec = np.array([[1,2]]).T
print(mat.dot(vec))   # The answer will be three rows and one column
print(vec.T.dot(vec)) # The answer will be a matrix with one row and one column. Note that it is a two dimensional array, not a scalar value

[[ 5]
 [11]
 [17]]
[[5]]


<!-- JPN -->
　もしベクトルの内積をスカラー値として取り出したい場合は、以下のように明示的に値を取り出そう。

<!-- ENG -->
　If you want to extract the inner product of vectors as a scalar value, extract the value explicitly as follows.

In [35]:
vec.T.dot(vec)[0,0] # Extract an element at row 0 and column 0 from the calculation results

5

-----

<!-- JPN -->
##### 課題 2

$A$ および $\boldsymbol{b}$ を入力とし、 $A^T b$ を計算する関数 `multiply1(mat_a, vec_b)` を実装せよ。

<!-- ENG -->
##### Exercise 2
　Implement `multiply1(mat_a, vec_b)` which outputs $A^T b$ with $A$ and $\boldsymbol{b}$ as inputs.

In [36]:
# Write your answers here
def multiply1(mat_a, vec_b):
  result = np.dot(mat_a.T, vec_b)
  return result

--------------

<!-- JPN -->
##### 課題 3

$X$ および $Y$ を入力とし、 $(X Y)^T$ を計算する関数 `multiply2(x, y)` を実装せよ。**ただし、 `x` および `y` は行列であるときもベクトルであるときも `multiply2()` 関数が正常に動作するようにすること。**

<!-- ENG -->
##### Exercise 3
　Implement `multiply2(x, y)` which outputs $(X Y)^T$ with $X$ and $Y$ as inputs. **Note that the `multiply2` function should work correctly when `x` and `y` are both matrices or vectors.**

In [37]:
# Write your answers here
def multiply2(x, y):
  result = np.dot(x, y)
  if isinstance(result, np.ndarray):
    return result.T
  else:
    return result

-------

<!-- JPN -->
### 2.4 | 逆行列の計算

　逆行列の計算は `np.linalg.inv()` 関数を用いる。linalgとは、linear algebra （線形代数）の略であり、複雑な演算が多数定義されている。

<!-- ENG -->
### 2.4 | Inverse matrix calculations

　Use the `np.linalg.inv()` function to calculate the inverse matrix. linalg is an abbreviation for linear algebra, which defines many complex operations.

In [38]:
mat1 = np.array([[1,2],
                 [3,4]]) # Regular matrix

In [39]:
print(np.linalg.inv(mat1))
print(np.linalg.inv(mat1).dot(mat1))  # Check that it is an inverse matrix

[[-2.   1. ]
 [ 1.5 -0.5]]
[[1.00000000e+00 0.00000000e+00]
 [1.11022302e-16 1.00000000e+00]]


<!-- JPN -->
　Pythonの実数値は有限の桁数で計算を行うため、厳密には逆行列との行列積は単位行列にはならなかったが、 $1.11 \times 10^{-16}$ は極めて小さな値なので0と見なして問題ない。このようなことはプログラミングにおける数値計算ではよくあることなので、**実数値を扱う時には微妙に値がずれることがある**ことを覚えておこう。

　**逆行列の計算は次回の基盤人工知能演習で利用する**など、機械学習との関連性が深いため、後述の固有値・固有ベクトルの演算と共に、データサイエンスや機械学習を行う上で必須の知識と言ってよい。

<!-- ENG -->
　The matrix product of the original matrix and its inverse matrix is not strictly an identity matrix because real numbers in Python are calculated with a finite number of digits, but since $1.11\times 10^{-16}$ is an extremely small value, it can be safely assumed to be zero. This is a common occurrence in numerical calculations in programming, so keep in mind that **the values may be slightly different when dealing with real numbers**.

　**Inverse matrix calculations are closely related to machine learning, and will be used in the next Exercises in Fundamental Artificial Intelligence**, so it is essential knowledge for data science and machine learning, along with eigenvalue and eigenvector operations that are described later.

<!-- JPN -->
　ところで、逆行列が存在しないようなケースで、この関数はどういう挙動を示すだろうか。「正則でない正方行列」「正方でない行列」で試してみると、（正しく）エラーを発生させることが確認できる。

<!-- ENG -->
　By the way, how would this function behave in a case where there is no inverse matrix? If you try it with a "singular matrix" and "non-square matrix", you will see that it generates errors (correctly).

In [40]:
mat2 = np.ones((2,2))       # Create a matrix filled with 2 rows and 2 columns of 1's, but this is a square matrix that is not regular
print(mat2)
print(np.linalg.inv(mat2))  # LinAlgError is generated because it is not regular

[[1. 1.]
 [1. 1.]]


LinAlgError: ignored

In [41]:
mat3 = np.ones((2, 3))      # Non-square matrix
print(mat3)
print(np.linalg.inv(mat3))  # LinAlgError is generated because it is not a square matrix

[[1. 1. 1.]
 [1. 1. 1.]]


LinAlgError: ignored

<!-- JPN -->
### 2.5 | 固有値分解：固有値・固有ベクトルの算出
　次に、`np.linalg.eig()`を利用して、正方行列$A$の固有値・固有ベクトルを求めてみる。固有値、固有ベクトルとは、正方行列$A$に対して、$A \boldsymbol{x} = \lambda \boldsymbol{x}$ を満たすスカラー値 $\lambda$ とベクトル $\boldsymbol{x} (\boldsymbol{x}\neq\boldsymbol{0})$ のことを指す。


<!-- ENG -->
### 2.5 | Eigendecomposition: computing eigenvalues and eigenvectors
　Next, we will use `np.linalg.eig()` to find the eigenvalues and eigenvectors of the square matrix $A$. The eigenvalues ​​and eigenvectors refer to the scalar value $\lambda$ and the vector $\boldsymbol{x} (\boldsymbol{x}\neq\boldsymbol{0})$ that satisfy $A \boldsymbol{x} = \lambda \boldsymbol{x}$ for the square matrix $A$.


In [42]:
mat = np.array([[1,2],
                [2,1]])
lambdas, xs = np.linalg.eig(mat)
print(lambdas[0], xs[:,0])
print(lambdas[1], xs[:,1])

3.0000000000000004 [0.70710678 0.70710678]
-0.9999999999999996 [-0.70710678  0.70710678]


<!-- JPN -->
　逆行列を求めた時と同様に微妙な誤差が発生しているが、以下の2つの等号が成り立っているようである。

<!-- ENG -->
　As with finding the inverse matrix, there are subtle errors, but it seems that the following two equations are satisfied.

<!-- BOTH -->
$$
\begin{aligned}
A \left[\frac{1}{\sqrt{2}}, \frac{1}{\sqrt{2}} \right]^T &= 3 \left[\frac{1}{\sqrt{2}}, \frac{1}{\sqrt{2}} \right]^T \\
A \left[-\frac{1}{\sqrt{2}}, \frac{1}{\sqrt{2}} \right]^T &= -1 \left[-\frac{1}{\sqrt{2}}, \frac{1}{\sqrt{2}} \right]^T
\end{aligned}
$$

<!-- JPN -->
　一応、実際にこの等号が成り立っているか確認してみよう。

<!-- ENG -->
　Let's see if these equations are actually satisfied.

In [43]:
print(mat.dot(xs[:,0]))     # Left side of the first equation
print(lambdas[0] * xs[:,0]) # Right side of the first equation

print("----------")

print(mat.dot(xs[:,1]))     # Left side of the second equation
print(lambdas[1] * xs[:,1]) # Right side of the second equation

[2.12132034 2.12132034]
[2.12132034 2.12132034]
----------
[ 0.70710678 -0.70710678]
[ 0.70710678 -0.70710678]


In [44]:
# It can actually be calculated together
# The results of each calculation are obtained for each column
print(mat.dot(xs))
print(lambdas * xs)

[[ 2.12132034  0.70710678]
 [ 2.12132034 -0.70710678]]
[[ 2.12132034  0.70710678]
 [ 2.12132034 -0.70710678]]


<!-- JPN -->
　出力される固有ベクトルは**列ベクトルが複数横に並べられた形**になっていることに注意して欲しい。

　固有値や固有ベクトルは、（基盤人工知能ではなく）基盤データサイエンスで講義が予定されている**「主成分分析 (Principal Component Analysis; PCA) 」**などの手法で利用されるので、しっかりとおさえておきたい。

<!-- ENG -->
　Note that the eigenvectors that are output are **in the form of multiple column vectors arranged side by side**.

　Eigenvalues and eigenvectors are used in methods such as **Principal Component Analysis (PCA)**, which is planned to be taught in Fundamental Data Science (not Fundamental Artificial Intelligence), so it is important to be familiar with them.

------

<!-- JPN -->
##### 課題 4

疑似逆行列 $(X^TX)^{-1}X^T$ を求める関数 `pseudo_inv(X)` を実装せよ。

<!-- ENG -->
##### Exercise 4
　Implement the function `pseudo_inv(X)` which computes the pseudo-inverse matrix $(X^TX)^{-1}X^T$.

In [45]:
# Write your answers here
def pseudo_inv(X):
  pseudo_inverse = np.linalg.pinv(X)
  return pseudo_inverse

------

<!-- JPN -->
##### 課題 5

　行列のサイズが $N \times N$ であるとする。$N$が変化するとき、**逆行列の計算に必要な時間はどのように変化するだろうか？ 計算時間がおおよそ $N$ の何乗に比例するかを整数で答えよ**（ヒント： $N=2000$ の場合に1秒以下、$N=8000$ の場合には50秒程度の計算時間を要するので、その周辺で数回計算を行い、計算速度を推定するのが良い）。この課題については、各自のPCで実行するとウェブブラウザや他のプログラムに影響されてしまうため、**Google Colab上で実行せよ**。

　時間計測は以下のコードで行うことができるので、適宜利用せよ。


<!-- ENG -->
##### Exercise 5

　Suppose the size of the matrix is $N \times N$. When $N$ varies, **how is the time required to calculate the inverse matrix changed? Answer an integer $k$ when calculation time $t$ is approximately proportional to $k$-th power of $N$**. (Hint: It takes less than 1 second for $N=2000$, and about 50 seconds for $N=8000$, so it's better to do a few calculations around there and estimate the calculation speed.) **This exercise must be done with Google Colab** since web browsers or other applications affect the results on your own PC.

　Time measurement can be performed with the following code, so use it accordingly.


In [47]:
import time
import numpy as np
ident = np.eye(8000) # np.eye(k): Create an identity matrix of k rows and k columns

st = time.time() # Measurement of start time
ident_inv = np.linalg.inv(ident)
ed = time.time() # Measurement of end time
print(ed - st) # Calculate the difference in seconds

33.71249866485596


In [50]:
import time
import numpy as np
ident = np.eye(2000) # np.eye(k): Create an identity matrix of k rows and k columns

st = time.time() # Measurement of start time
ident_inv = np.linalg.inv(ident)
ed = time.time() # Measurement of end time
print(ed - st) # Calculate the difference in seconds

0.5713083744049072


<!-- JPN -->
　なお、計算を行うためにかかる時間のことを**時間計算量 (time complexity)** と呼び、計算すべきデータのサイズ $N$ に対して、$N^2$ に比例する時間か、それ未満で計算できる場合は **時間計算量は $O(N^2)$ である** などと表現する（$O(\cdot)$ をビッグオー記法と言う）。

<!-- ENG -->
　The time it takes to perform a calculation is called **time complexity**, and if the calculation can be performed in a time proportional to or less than $N^2$ for a data size $N$ to be calculated, it is described that **the time complexity is $O(N^2)$** ($O(\cdot)$ is referred to as Big O notation).

<!-- JPN -->
<font color=orange> **このテキストセルに答案を記述せよ。** </font>

<!-- ENG -->
<font color=orange> **Write your answer in this text cell.** </font>

The ratio of time taken for $N = 8000$ (33.71s) to $N = 2000$ (0.57s) is approximately $33.71/0.57 = 59.14$.

$8000/2000 = 4$, indicating a factor of 4 increase in $N$.

So, $k \approx \log _{4}(59.14) \approx 2.94$.

Rounded to the nearest integer, $k \approx 3$.

Hence, the calculation time $t$ is approximately proportional to the $t$-th power of $N$, where $k = 3$.




---------

<!-- JPN -->
## 補足資料




<!-- ENG -->
## Supplementary Material




<!-- JPN -->
### ※1 `np.array`と`np.matrix`
　線形代数を行う上では要素積よりも行列積を行うことが多い。このような場合には `np.matrix` を用いると良く、これは `A * B` と記述するだけで `A`と`B`の行列積を算出してくれる。一方、本演習では、線形代数の文脈においても `np.array` を用いることとしている。これには以下のような理由がある。

- 線形代数を行う以外にも、機械学習を行うためのデータの管理等でNumPyを使う。この場合は要素積を取る方が都合の良いケースがある。
- 初めてNumPyに触れる受講生に`np.array` と `np.matrix` の使い分けを要求することは困難である。
- `np.array` を用いた実例が世の中にあふれており、自学自習がより容易である。

<!-- ENG -->
### S1 `np.array` and `np.matrix`
　In doing linear algebra, we often do matrix products rather than element products. In such a case, you can use `np.matrix`, which will calculate the matrix product of `A` and `B` by simply writing `A * B`. However, in this exercise, we also use `np.array` in the context of linear algebra. There are several reasons for this.

- In addition to doing linear algebra, NumPy is used to manage data, etc. for machine learning. Given this, there are cases in which it is more convenient to take the element product.
- It is difficult to ask students who are new to NumPy to distinguish between `np.array` and `np.matrix`.
- There are a lot of examples in the world where `np.arrays` can be used, and it makes self-study much easier.

<!-- JPN -->
### ※2 `np.array()`の正体
　本文では「関数のようなもの」と表現したが、実際には`np.array`は**クラス**であり、`np.array()`クラスの**コンストラクタ**、それによって作られる`vec`や`mat`などは**インスタンス**と呼ばれるものである。

　このあたりについては、「オブジェクト指向」という単語でGoogle検索を行うと、理解が深まるかもしれない。

<!-- ENG -->
### S2 What an `np.array ()` is
　In this course text, we described it as "something like a function", but in reality, `np.array` is a **class**, and the **constructors** of the `np.array()` class and the `vec`, `mat`, etc. created by it are called **instances**.

　For a better understanding of this, a Google search on the word "object-oriented" may help.

<!-- JPN -->
### ※3 ブロードキャストについて
　本文中で記載したブロードキャストが行われる条件は正確には不十分であり、ChainerのNumPy入門の説明を借りると「**2つの配列の各次元が同じ大きさになっているか、どちらかが1である**」ときにブロードキャストが可能である。

<!-- ENG -->
### S3 Broadcasting
　The conditions for broadcast described in the text are not exactly sufficient. To borrow from the Intro to NumPy in Chainer Tutorial, broadcast is possible when **"each dimension of the two arrays has the same size, if not one of them should be 1"**.

In [51]:
Ma = np.array([[1, 2], [3, 4], [5, 6]])
Mb = np.array([[7, 8]])
Mc = np.array([[9],[10]])
print(Ma.shape)
print(Mb.shape)
print(Mc.shape)

(3, 2)
(1, 2)
(2, 1)


In [52]:
print(Ma + Mb) # (3,2) and (1,2) satisfy the conditions and can be calculated

[[ 8 10]
 [10 12]
 [12 14]]


In [53]:
print(Ma + Mc) # Error occurred because (3,2) and (2,1) do not satisfy the conditions

ValueError: ignored

<!-- JPN -->
　2次元配列と1次元配列の間のブロードキャストの場合、**次元数の少ないベクトル側を`.reshape()`で行列にして**、同じ演算を行う。

<!-- ENG -->
　In the case of a broadcast between a two dimensional array and a one dimensional array, perform the same operation **by making the vector that has fewer dimensions into a matrix using `.reshape()`**.

In [54]:
M = np.array([[1, 2, 3], [4, 5, 6]])
v = np.array([7, 8, 9])
print(M.shape)
print(v.shape)

(2, 3)
(3,)


In [55]:
print(M + v)
print(M + v.reshape(1,-1)) # Same meaning

[[ 8 10 12]
 [11 13 15]]
[[ 8 10 12]
 [11 13 15]]


<!-- JPN -->
　より詳しい情報は[公式リファレンス](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html)を参照してほしい。

<!-- ENG -->
　For more detailed information, please refer to [the official reference](https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html).